Tests/SwiftGRPCNIOTests/NIOServerTests.swift

```diff
@@ -92,211 +94,106 @@ class NIOServerTests: NIOServerTestCase {
-      XCTAssertEqual(.ok, callResult.statusCode)
-      completionHandlerExpectation.fulfill()
-    }
-
-    for string in NIOServerTests.lotsOfStrings {
-      XCTAssertEqual("Swift echo expand (\(string)): \(string)", try! call.receive()!.text)
-    }
-    XCTAssertNil(try! call.receive())
-
-    waitForExpectations(timeout: defaultTimeout)
+    XCTAssertNoThrow(try doTestServerStreaming(messages: NIOServerTests.lotsOfStrings))
   }
 }
 
 extension NIOServerTests {
-  func testBidirectionalStreamingBatched() {
-    let finalCompletionHandlerExpectation = expectation(description: "final completion handler called")
-    let call = try! client.update { callResult in
-      XCTAssertEqual(.ok, callResult.statusCode)
-      finalCompletionHandlerExpectation.fulfill()
+  private func doTestBidirectionalStreaming(messages: [String], waitForEachResponse: Bool = false) throws {
+    let responseReceived = waitForEachResponse ? DispatchSemaphore(value: 0) : nil
+    var index = 0
+
+    let call = client.update { response in
+      XCTAssertEqual("Swift echo update (\(index)): \(messages[index])", response.text)
+      responseReceived?.signal()
+      index += 1
     }
 
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "foo")))
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "bar")))
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "baz")))
-
-    call.waitForSendOperationsToFinish()
-
-    XCTAssertNoThrow(try call.closeSend())
-
-    XCTAssertEqual("Swift echo update (0): foo", try! call.receive()!.text)
-    XCTAssertEqual("Swift echo update (1): bar", try! call.receive()!.text)
-    XCTAssertEqual("Swift echo update (2): baz", try! call.receive()!.text)
-    XCTAssertNil(try! call.receive())
-
-    waitForExpectations(timeout: defaultTimeout)
-  }
-
-  func testBidirectionalStreamingPingPong() {
-    let finalCompletionHandlerExpectation = expectation(description: "final completion handler called")
-    let call = try! client.update { callResult in
-      XCTAssertEqual(.ok, callResult.statusCode)
-      finalCompletionHandlerExpectation.fulfill()
+    messages.forEach { part in
+      call.send(.message(Echo_EchoRequest.with { $0.text = part }))
+      responseReceived?.wait()
     }
+    call.send(.end)
 
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "foo")))
-    XCTAssertEqual("Swift echo update (0): foo", try! call.receive()!.text)
-
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "bar")))
-    XCTAssertEqual("Swift echo update (1): bar", try! call.receive()!.text)
-
-    XCTAssertNoThrow(try call.send(Echo_EchoRequest(text: "baz")))
-    XCTAssertEqual("Swift echo update (2): baz", try! call.receive()!.text)
-
-    call.waitForSendOperationsToFinish()
-
-    XCTAssertNoThrow(try call.closeSend())
-
-    XCTAssertNil(try! call.receive())
-
-    waitForExpectations(timeout: defaultTimeout)
+    XCTAssertEqual(try call.status.wait().code, .ok)
```

Comment:
Also assert that `index == messages.count` here, or that all expectations are fulfilled (with one expectation for each sent message)?